File: LeetCode/427ConstructQuadTree/ConstructQuadTree.py
class Solution:
    def same_values(self, grid):
        val = grid[0][0]
        for i in range(len(grid)):
            for j in range(len(grid)):
                if grid[i][j] != val:
                    return False
        return True
    def slice_grid(self, grid, i, j, k, l):
        new_grid = grid[i:j]
        for v in range(len(new_grid)):
            new_grid[v] = new_grid[v][k:l]
        return new_grid
    def construct(self, grid: List[List[int]]) -> 'Node':
        # Divide and Conquer -> Time:O(n*log(n)), Space:O(log(n))
        if self.same_values(grid):
            return Node(grid[0][0], 1, None, None, None, None)
        n = len(grid)
        new_node = Node(grid[0][0], 0, None, None, None, None)
        new_node.topLeft = self.construct(self.slice_grid(grid, 0, n//2, 0, n//2))
        new_node.topRight = self.construct(self.slice_grid(grid, 0, n//2, n//2, n))
        new_node.bottomLeft = self.construct(self.slice_grid(grid, n//2, n, 0, n//2))
        new_node.bottomRight = self.construct(self.slice_grid(grid, n//2, n, n//2, n))
        return new_node

File: LeetCode/427ConstructQuadTree/ConstructQuadTree.cpp
class Solution {
public:
    bool sameValues(vector<vector<int>>& grid) {
        int val = grid[0][0];
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid.size(); j++) {
                if (grid[i][j] != val) {
                    return false;
                }
            }
        }
        return true;
    }
    vector<vector<int>> sliceQuarter(vector<vector<int>>& grid, int i, int j, int k, int l) {
        vector<vector<int>> result;
        vector<int> temp;
        for (int a = i; a < j; a++) {
            temp.clear();
            for (int b = k; b < l; b++) {
                temp.push_back(grid[a][b]);
            }
            result.push_back(temp);
        }
        return result;
    }
    Node* construct(vector<vector<int>>& grid) {
        // Divide and Conquer -> Time:O(n*log(n)), Space:O(log(n))
        if (sameValues(grid)) {
            return new Node(bool(grid[0][0]), true);
        }
        Node* newNode = new Node();
        int n = grid.size();
        vector<vector<int>> quarterGrid = sliceQuarter(grid, 0, n/2, 0, n/2);
        newNode->topLeft = construct(quarterGrid);
        quarterGrid = sliceQuarter(grid, 0, n/2, n/2, n);
        newNode->topRight = construct(quarterGrid);
        quarterGrid = sliceQuarter(grid, n/2, n, 0, n/2);
        newNode->bottomLeft = construct(quarterGrid);
        quarterGrid = sliceQuarter(grid, n/2, n, n/2, n);
        newNode->bottomRight = construct(quarterGrid);
        return newNode;
    }
};
